src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1453,42 +1453,120 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                     }
                 }
 
-                DiagnosticBag warnings = DiagnosticBag.GetInstance();
-                var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
+                var intOrVoidEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
+                // Validity and diagnostics are also tracked because they must be conditionally handled 
+                // if there are not any "traditional" entrypoints found.
+                var taskEntryPoints = ArrayBuilder<(bool IsValid, MethodSymbol Candidate, DiagnosticBag SpecificDiagnostics)>.GetInstance();
+
+                // These diagnostics (warning only) are added to the compilation only if 
+                // there were not any main methods found.
+                DiagnosticBag noMainFoundDiagnostics = DiagnosticBag.GetInstance();
+
                 foreach (var candidate in entryPointCandidates)
                 {
-                    if (!HasEntryPointSignature(candidate, warnings))
+                    var perCandidateBag = DiagnosticBag.GetInstance();
+                    var (IsCandidate, IsTaskLike) = HasEntryPointSignature(candidate, perCandidateBag);
+
+                    if (IsTaskLike)
                     {
-                        // a single error for partial methods
-                        warnings.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
-                        continue;
+                        taskEntryPoints.Add((IsCandidate, candidate, perCandidateBag));
                     }
+                    else
+                    {
+                        if (IsCandidate)
+                        {
+                            intOrVoidEntryPoints.Add(candidate);
+                            Debug.Assert(perCandidateBag.IsEmptyWithoutResolution);
+                        }
+                        else
+                        {
+                            noMainFoundDiagnostics.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
+                            noMainFoundDiagnostics.AddRange(perCandidateBag);
+                        }
+                        perCandidateBag.Free();
+                    }
+                }
+
+                var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
 
+                bool CheckGenericity(MethodSymbol candidate)
```

Comment:
Consider including `IsValid` check as well (to report `ErrorCode.WRN_InvalidMainSig`) since that same check appears in both cases where this local function is called.